File: packages/runtime/src/schedule-materializer.ts
// @ts-nocheck -- Vendored TanStack source targets a looser optional-property TypeScript policy.
import type {
  ScheduleId,
  WorkflowRegistrationMap,
  WorkflowRuntimeDefinition,
  WorkflowScheduleDefinition,
  WorkflowScheduleSpec
} from "./types"

const DEFAULT_CRON_LOOKBACK_MS = 32 * 24 * 60 * 60 * 1000

export interface MaterializeWorkflowSchedulesOptions {
  now?: number
  cronLookbackMs?: number
}

export type MaterializedWorkflowSchedule =
  | {
    kind: "materialized"
    workflowId: string
    scheduleId: ScheduleId
    fireAt: number
    schedule: WorkflowScheduleSpec
  }
  | {
    kind: "disabled"
    workflowId: string
    scheduleId: ScheduleId
    schedule: WorkflowScheduleSpec
  }
  | {
    kind: "not-due"
    workflowId: string
    scheduleId: ScheduleId
    schedule: WorkflowScheduleSpec
  }

export async function materializeWorkflowSchedules<
  TWorkflows extends WorkflowRegistrationMap
>(
  runtime: WorkflowRuntimeDefinition<TWorkflows>,
  options: MaterializeWorkflowSchedulesOptions = {}
): Promise<Array<MaterializedWorkflowSchedule>> {
  const now = options.now ?? Date.now()
  const cronLookbackMs = options.cronLookbackMs ?? DEFAULT_CRON_LOOKBACK_MS
  const materialized: Array<MaterializedWorkflowSchedule> = []

  if (!Number.isFinite(cronLookbackMs) || cronLookbackMs < 0) {
    throw new Error("Workflow cron lookback must be a non-negative number.")
  }

  for (const [workflowId, registration] of Object.entries(runtime.workflows)) {
    const schedules = registration.schedules ?? []
    for (let index = 0; index < schedules.length; index++) {
      const definition = schedules[index]!
      const scheduleId = getScheduleId(workflowId, definition, index)

      if (definition.enabled === false) {
        await runtime.store.upsertSchedule({
          scheduleId,
          workflowId,
          workflowVersion: registration.version,
          schedule: definition.schedule,
          overlapPolicy: definition.overlapPolicy ?? "skip",
          input: undefined,
          nextFireAt: undefined,
          enabled: false,
          now
        })
        materialized.push({
          kind: "disabled",
          workflowId,
          scheduleId,
          schedule: definition.schedule
        })
        continue
      }

      const fireAt = getDueFireAt(definition.schedule, now, cronLookbackMs)
      if (fireAt === undefined) {
        materialized.push({
          kind: "not-due",
          workflowId,
          scheduleId,
          schedule: definition.schedule
        })
        continue
      }

      await runtime.store.upsertSchedule({
        scheduleId,
        workflowId,
        workflowVersion: registration.version,
        schedule: definition.schedule,
        overlapPolicy: definition.overlapPolicy ?? "skip",
        input: await resolveScheduleInput(definition.input),
        nextFireAt: fireAt,
        enabled: true,
        now
      })
      materialized.push({
        kind: "materialized",
        workflowId,
        scheduleId,
        fireAt,
        schedule: definition.schedule
      })
    }
  }

  return materialized
}

function getScheduleId(
  workflowId: string,
  definition: WorkflowScheduleDefinition,
  index: number
): ScheduleId {
  return definition.id ?? `${workflowId}:${index}`
}

async function resolveScheduleInput(
  input: WorkflowScheduleDefinition["input"]
) {
  return typeof input === "function" ? await input() : input
}

function getDueFireAt(
  schedule: WorkflowScheduleSpec,
  now: number,
  cronLookbackMs: number
) {
  if (schedule.kind === "interval") {
    if (!Number.isFinite(schedule.everyMs) || schedule.everyMs <= 0) {
      throw new Error(
        "Interval workflow schedules must use a positive everyMs."
      )
    }
    return Math.floor(now / schedule.everyMs) * schedule.everyMs
  }

  return getPreviousCronFireAt(schedule, now, cronLookbackMs)
}

function getPreviousCronFireAt(
  schedule: Extract<WorkflowScheduleSpec, { kind: "cron" }>,
  now: number,
  lookbackMs: number
) {
  if (schedule.timezone && schedule.timezone !== "UTC") {
    throw new Error(
      `Workflow cron schedules are materialized in UTC. Received timezone "${schedule.timezone}".`
    )
  }

  const cron = parseCronExpression(schedule.expression)
  const start = floorToMinute(now)
  const end = start - lookbackMs

  for (let timestamp = start; timestamp >= end; timestamp -= 60_000) {
    if (matchesCron(cron, new Date(timestamp))) return timestamp
  }

  return undefined
}

interface ParsedCronExpression {
  minute: ParsedCronField
  hour: ParsedCronField
  dayOfMonth: ParsedCronField
  month: ParsedCronField
  dayOfWeek: ParsedCronField
}

interface ParsedCronField {
  wildcard: boolean
  values: ReadonlySet<number>
}

function parseCronExpression(expression: string): ParsedCronExpression {
  const fields = expression.trim().split(/\s+/)
  if (fields.length !== 5) {
    throw new Error(
      `Workflow cron schedules must use five fields. Received "${expression}".`
    )
  }

  return {
    minute: parseCronField(fields[0]!, 0, 59),
    hour: parseCronField(fields[1]!, 0, 23),
    dayOfMonth: parseCronField(fields[2]!, 1, 31),
    month: parseCronField(fields[3]!, 1, 12),
    dayOfWeek: parseCronField(fields[4]!, 0, 7, normalizeDayOfWeek)
  }
}

function parseCronField(
  field: string,
  min: number,
  max: number,
  normalize: (value: number) => number = (value) => value
): ParsedCronField {
  const values = new Set<number>()
  const parts = field.split(",")

  for (const part of parts) {
    const [rangePart, stepPart] = part.split("/")
    const step = stepPart === undefined ? 1 : Number(stepPart)
    if (!Number.isInteger(step) || step <= 0) {
      throw new Error(`Invalid cron step "${part}".`)
    }

    const range = parseCronRange(rangePart!, min, max)
    for (let value = range.start; value <= range.end; value += step) {
      values.add(normalize(value))
    }
  }

  return {
    wildcard: field === "*",
    values
  }
}

function parseCronRange(range: string, min: number, max: number) {
  if (range === "*") return { start: min, end: max }

  const bounds = range.split("-")
  if (bounds.length === 1) {
    const value = parseCronNumber(bounds[0]!, min, max)
    return { start: value, end: value }
  }
  if (bounds.length === 2) {
    const start = parseCronNumber(bounds[0]!, min, max)
    const end = parseCronNumber(bounds[1]!, min, max)
    if (end < start) throw new Error(`Invalid cron range "${range}".`)
    return { start, end }
  }

  throw new Error(`Invalid cron range "${range}".`)
}

function parseCronNumber(value: string, min: number, max: number) {
  const parsed = Number(value)
  if (!Number.isInteger(parsed) || parsed < min || parsed > max) {
    throw new Error(`Invalid cron value "${value}".`)
  }
  return parsed
}

function normalizeDayOfWeek(value: number) {
  return value === 7 ? 0 : value
}

function matchesCron(cron: ParsedCronExpression, date: Date) {
  const dayOfMonthMatches = cron.dayOfMonth.values.has(date.getUTCDate())
  const dayOfWeekMatches = cron.dayOfWeek.values.has(date.getUTCDay())
  const dayMatches = !cron.dayOfMonth.wildcard && !cron.dayOfWeek.wildcard
    ? dayOfMonthMatches || dayOfWeekMatches
    : dayOfMonthMatches && dayOfWeekMatches

  return (
    cron.minute.values.has(date.getUTCMinutes()) &&
    cron.hour.values.has(date.getUTCHours()) &&
    dayMatches &&
    cron.month.values.has(date.getUTCMonth() + 1)
  )
}

function floorToMinute(timestamp: number) {
  return Math.floor(timestamp / 60_000) * 60_000
}

File: src/Firegrid.Runtime/ScheduleMaterializer.fs
namespace Firegrid.Runtime

open System.Collections.Generic
open Fable.Core
open Firegrid.Core

/// `schedule-materializer.ts` — cron parser + `materializeWorkflowSchedules`.
[<RequireQualifiedAccess>]
module ScheduleMaterializer =

    let private DEFAULT_CRON_LOOKBACK_MS = 32.0 * 24.0 * 60.0 * 60.0 * 1000.0

    type MaterializeWorkflowSchedulesOptions =
        { Now: float option
          CronLookbackMs: float option }

        static member Empty = { Now = None; CronLookbackMs = None }

    /// `MaterializedWorkflowSchedule` — discriminated on `kind`.
    type MaterializedWorkflowSchedule =
        | Materialized of workflowId: string * scheduleId: ScheduleId * fireAt: float * schedule: WorkflowScheduleSpec
        | Disabled of workflowId: string * scheduleId: ScheduleId * schedule: WorkflowScheduleSpec
        | NotDue of workflowId: string * scheduleId: ScheduleId * schedule: WorkflowScheduleSpec

    // ── cron parsing ──────────────────────────────────────────────────

    type private ParsedCronField =
        { Wildcard: bool
          Values: HashSet<int> }

    type private ParsedCronExpression =
        { Minute: ParsedCronField
          Hour: ParsedCronField
          DayOfMonth: ParsedCronField
          Month: ParsedCronField
          DayOfWeek: ParsedCronField }

    type private CronRange = { Start: int; End: int }

    let private parseCronNumber (value: string) (min: int) (max: int) : int =
        let parsed = RuntimeSdk.numberValue (box value)

        if not (RuntimeSdk.numberIsInteger (box parsed)) || parsed < float min || parsed > float max then
            failwithf "Invalid cron value \"%s\"." value

        int parsed

    let private parseCronRange (range: string) (min: int) (max: int) : CronRange =
        if range = "*" then
            { Start = min; End = max }
        else
            let bounds = range.Split('-')

            if bounds.Length = 1 then
                let value = parseCronNumber bounds[0] min max
                { Start = value; End = value }
            elif bounds.Length = 2 then
                let start = parseCronNumber bounds[0] min max
                let endVal = parseCronNumber bounds[1] min max

                if endVal < start then
                    failwithf "Invalid cron range \"%s\"." range

                { Start = start; End = endVal }
            else
                failwithf "Invalid cron range \"%s\"." range

    let private parseCronField
        (field: string)
        (min: int)
        (max: int)
        (normalize: int -> int)
        : ParsedCronField =
        let values = HashSet<int>()
        let parts = field.Split(',')

        for part in parts do
            let slashParts = part.Split('/')
            let rangePart = slashParts[0]

            let step =
                if slashParts.Length < 2 then
                    1.0
                else
                    RuntimeSdk.numberValue (box slashParts[1])

            if not (RuntimeSdk.numberIsInteger (box step)) || step <= 0.0 then
                failwithf "Invalid cron step \"%s\"." part

            let range = parseCronRange rangePart min max
            let mutable value = range.Start

            while value <= range.End do
                values.Add(normalize value) |> ignore
                value <- value + int step

        { Wildcard = field = "*"
          Values = values }

    let private normalizeDayOfWeek (value: int) : int = if value = 7 then 0 else value

    let private identity (value: int) : int = value

    let private parseCronExpression (expression: string) : ParsedCronExpression =
        let fields = RuntimeSdk.splitRegex (RuntimeSdk.trim expression) "\\s+"

        if fields.Length <> 5 then
            failwithf "Workflow cron schedules must use five fields. Received \"%s\"." expression

        { Minute = parseCronField fields[0] 0 59 identity
          Hour = parseCronField fields[1] 0 23 identity
          DayOfMonth = parseCronField fields[2] 1 31 identity
          Month = parseCronField fields[3] 1 12 identity
          DayOfWeek = parseCronField fields[4] 0 7 normalizeDayOfWeek }

    let private matchesCron (cron: ParsedCronExpression) (date: RuntimeSdk.JsDate) : bool =
        let dayOfMonthMatches = cron.DayOfMonth.Values.Contains(RuntimeSdk.getUTCDate date)
        let dayOfWeekMatches = cron.DayOfWeek.Values.Contains(RuntimeSdk.getUTCDay date)

        let dayMatches =
            if not cron.DayOfMonth.Wildcard && not cron.DayOfWeek.Wildcard then
                dayOfMonthMatches || dayOfWeekMatches
            else
                dayOfMonthMatches && dayOfWeekMatches

        cron.Minute.Values.Contains(RuntimeSdk.getUTCMinutes date)
        && cron.Hour.Values.Contains(RuntimeSdk.getUTCHours date)
        && dayMatches
        && cron.Month.Values.Contains(RuntimeSdk.getUTCMonth date + 1)

    let private floorToMinute (timestamp: float) : float =
        RuntimeSdk.mathFloor (timestamp / 60000.0) * 60000.0

    let private getPreviousCronFireAt
        (expression: string)
        (timezone: string option)
        (now: float)
        (lookbackMs: float)
        : float option =
        match timezone with
        | Some tz when tz <> "UTC" ->
            failwithf "Workflow cron schedules are materialized in UTC. Received timezone \"%s\"." tz
        | _ -> ()

        let cron = parseCronExpression expression
        let start = floorToMinute now
        let endTs = start - lookbackMs

        let mutable timestamp = start
        let mutable result = None

        while result.IsNone && timestamp >= endTs do
            if matchesCron cron (RuntimeSdk.newDate timestamp) then
                result <- Some timestamp
            else
                timestamp <- timestamp - 60000.0

        result

    let private getDueFireAt (schedule: WorkflowScheduleSpec) (now: float) (cronLookbackMs: float) : float option =
        match schedule with
        | IntervalSpec(everyMs, _) ->
            if not (RuntimeSdk.numberIsFinite (box everyMs)) || everyMs <= 0.0 then
                failwith "Interval workflow schedules must use a positive everyMs."

            Some(RuntimeSdk.mathFloor (now / everyMs) * everyMs)
        | CronSpec(expression, timezone) -> getPreviousCronFireAt expression timezone now cronLookbackMs

    // ── materialization ───────────────────────────────────────────────

    let private getScheduleId (workflowId: string) (definition: WorkflowScheduleDefinition) (index: int) : ScheduleId =
        match definition.Id with
        | Some id -> id
        | None -> sprintf "%s:%d" workflowId index

    let private resolveScheduleInput (input: obj option) : JS.Promise<obj> =
        promise {
            match input with
            | None -> return RuntimeSdk.undefinedValue
            | Some v ->
                if RuntimeSdk.isTypeofFunction v then
                    let fn = v :?> (unit -> obj)
                    let! resolved = RuntimeSdk.promiseResolve (fn ())
                    return resolved
                else
                    return v
        }

    /// `materializeWorkflowSchedules(runtime, options)`.
    let materialize
        (runtime: WorkflowRuntimeDefinition)
        (options: MaterializeWorkflowSchedulesOptions)
        : JS.Promise<MaterializedWorkflowSchedule[]> =
        promise {
            let now = options.Now |> Option.defaultValue (RuntimeSdk.nowMillis ())
            let cronLookbackMs = options.CronLookbackMs |> Option.defaultValue DEFAULT_CRON_LOOKBACK_MS
            let materialized = ResizeArray<MaterializedWorkflowSchedule>()

            if not (RuntimeSdk.numberIsFinite (box cronLookbackMs)) || cronLookbackMs < 0.0 then
                failwith "Workflow cron lookback must be a non-negative number."

            for (workflowId, registrationObj) in RuntimeSdk.objectEntries runtime.Workflows do
                let registration = registrationObj :?> WorkflowRegistration
                let schedules = registration.Schedules |> Option.defaultValue [||]

                for index in 0 .. schedules.Length - 1 do
                    let definition = schedules[index]
                    let scheduleId = getScheduleId workflowId definition index

                    if definition.Enabled = Some false then
                        do!
                            runtime.Store.UpsertSchedule
                                { ScheduleId = scheduleId
                                  WorkflowId = workflowId
                                  WorkflowVersion = registration.Version
                                  Schedule = definition.Schedule
                                  OverlapPolicy = definition.OverlapPolicy |> Option.defaultValue "skip"
                                  Input = None
                                  NextFireAt = None
                                  Enabled = false
                                  Now = now }

                        materialized.Add(Disabled(workflowId, scheduleId, definition.Schedule))
                    else
                        let fireAt = getDueFireAt definition.Schedule now cronLookbackMs

                        match fireAt with
                        | None -> materialized.Add(NotDue(workflowId, scheduleId, definition.Schedule))
                        | Some fire ->
                            let! resolvedInput = resolveScheduleInput definition.Input

                            do!
                                runtime.Store.UpsertSchedule
                                    { ScheduleId = scheduleId
                                      WorkflowId = workflowId
                                      WorkflowVersion = registration.Version
                                      Schedule = definition.Schedule
                                      OverlapPolicy = definition.OverlapPolicy |> Option.defaultValue "skip"
                                      Input = Some resolvedInput
                                      NextFireAt = Some fire
                                      Enabled = true
                                      Now = now }

                            materialized.Add(Materialized(workflowId, scheduleId, fire, definition.Schedule))

            return materialized.ToArray()
        }
